### PR TITLE
FastGCN

### PR DESCRIPTION
**Pull request by gpzlx1**
* Support FastGCN

@Liu-rj Profiling for fastgcn if possible
On my computer, the result is as follows
```shell
$ python ./examples/fastgcn.py
Check load successfully: [None, tensor([        0,         2,         4,  ..., 114844498, 114848855,
        114848857], device='cuda:0'), tensor([    15,      0,    151,  ..., 231024, 232963, 232964], device='cuda:0')] 

fastgcn sampling AVG: 2.2688998116387262  ms.
```

### CODE REVIEW

**Comment by Liu-rj** [comment]
There are bugs in `RowwiseSlicing`. Merging delayed.

**Comment by gpzlx1** [comment]
Support FastGCN[bias]

File: python/gs/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import torch
```
Comment (czkkkkkk) [inline python/gs/dataloader.py:1]: Could you make the dataloader a seperate PR so others can use it directly?

File: python/gs/matrix_api/matrix.py [hunk 2]
```diff
@@ -21,23 +23,33 @@ def load_dgl_graph(self, g):
         csc_indices = torch.tensor(csc.indices).long().cuda()
         self._graph.load_csc(csc_indptr, csc_indices)
 
+    def to_dgl_block(self):
```
Comment (czkkkkkk) [inline python/gs/matrix_api/matrix.py:26]: When is API used? I cannot find a reference in this PR.

File: python/gs/jit/optimize.py [hunk 3]
```diff
@@ -1,2 +1,119 @@
-def optimize(ir):
-    pass
\ No newline at end of file
+import operator
```
Comment (czkkkkkk) [inline python/gs/jit/optimize.py:1]: Looks like the optimization pass is also not very relevant to this FastGCN PR. Could you move it to another PR?

Comment (gpzlx1) [reply python/gs/jit/optimize.py:1]: done. Thery'are on https://github.com/czkkkkkk/graph_sampling/pull/15

File: src/cuda/list_sampling.cu [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * sampling for replace = True. It will return (selected_data, selected_index)
+ *
+ * @tparam IdType
+ * @param data
+ * @param num_picks
+ * @param replace
+ * @return std::tuple<torch::Tensor, torch::Tensor>
+ */
+template <typename IdType>
+std::tuple<torch::Tensor, torch::Tensor> _ListSampling(torch::Tensor data,
```
Comment (czkkkkkk) [inline src/cuda/list_sampling.cu:21]: As suggested by Gan Quan in the quip document, can we implement `ListSampling` by `torch.multimonial`?

**Review by czkkkkkk — CHANGES_REQUESTED** [review]

File: examples/fastgcn/load_graph.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import dgl
```
[the remark this answers is not in the document]
Comment (gpzlx1) [reply examples/fastgcn/load_graph.py:1]: @Liu-rj pls

File: examples/fastgcn/train_minibatch.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+                        *node.args[0].args,
+                        *node.args[1:],
+                    ))
+                node.replace_all_uses_with(new_node)
+    gm.graph.lint()
+    gm.recompile()
+    return gm
+
+
+class SAGE(nn.Module):
```
[the remark this answers is not in the document]
Comment (gpzlx1) [reply examples/fastgcn/train_minibatch.py:64]: @Liu-rj pls

File: src/cuda/list_sampling.cu (hunk 4 above)
Comment (gpzlx1) [reply src/cuda/list_sampling.cu:21]: However, `torch.multimonial` cannot support over 2^24 (16,777,216) items sampling on GPU.

File: examples/fastgcn/load_graph.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Liu-rj) [reply examples/fastgcn/load_graph.py:1]: OK. How about first merging this pr and then I open a new pr for this?

[the remark this answers is not in the document]
Comment (czkkkkkk) [reply examples/fastgcn/load_graph.py:1]: ok

**Review by czkkkkkk — APPROVED** [review]
